pkg/graveler/graveler.go

```diff
@@ -2528,6 +2530,124 @@ func (g *Graveler) Merge(ctx context.Context, repository *RepositoryRecord, dest
+		}
+		if !empty {
+			return nil, fmt.Errorf("%s: %w", destination, ErrDirtyBranch)
+		}
+		toCommit, err := g.dereferenceCommit(ctx, repository, Ref(destination))
+		if err != nil {
+			return nil, err
+		}
+
+		// Source is not a commit - merge base is destination root
+		itr, err := g.Log(ctx, repository, toCommit.CommitID, false)
+		if err != nil {
+			return nil, err
+		}
+		root := toCommit
+		for itr.Next() {
+			root = itr.Value()
+		}
+
+		g.log(ctx).WithFields(logging.Fields{
+			"repository":             repository.RepositoryID,
+			"source":                 source,
+			"destination":            destination,
+			"source_meta_range":      source,
+			"destination_meta_range": toCommit.MetaRangeID,
+			"base_meta_range":        root,
+		}).Trace("Merge")
+
+		metaRangeID, err := g.CommittedManager.MergeFromMetaRange(ctx, storageNamespace, toCommit.MetaRangeID, root.MetaRangeID, source, MergeStrategySrc)
+		if err != nil {
+			if !errors.Is(err, ErrUserVisible) {
+				err = fmt.Errorf("merge in CommitManager: %w", err)
+			}
+			return nil, err
+		}
+		commit = NewCommit()
+		commit.Committer = commitParams.Committer
+		commit.Message = commitParams.Message
+		commit.MetaRangeID = metaRangeID
+		commit.Parents = []CommitID{toCommit.CommitID}
+		commit.Generation = toCommit.Generation + 1
+		commit.Metadata = commitParams.Metadata
+		commit.Metadata[MergeStrategyMetadataKey] = MergeStrategySrcWinsStr
+		preRunID = g.hooks.NewRunID()
+		err = g.hooks.PreMergeHook(ctx, HookRecord{
+			EventType:        EventTypePreMerge,
+			RunID:            preRunID,
+			RepositoryID:     repository.RepositoryID,
+			StorageNamespace: storageNamespace,
+			BranchID:         destination,
+			SourceRef:        toCommit.CommitID.Ref(), // Will allow running pre-merge hooks of source ref
+			Commit:           commit,
+		})
+		if err != nil {
+			return nil, &HookAbortError{
+				EventType: EventTypePreMerge,
+				RunID:     preRunID,
+				Err:       err,
+			}
+		}
+		commitID, err = g.RefManager.AddCommit(ctx, repository, commit)
+		if err != nil {
+			return nil, fmt.Errorf("add commit: %w", err)
+		}
+
+		tokensToDrop = branch.SealedTokens
+		branch.SealedTokens = []StagingToken{}
+		branch.CommitID = commitID
+		return branch, nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("update branch %s: %w", destination, err)
+	}
+
+	g.dropTokens(ctx, tokensToDrop...)
+	postRunID := g.hooks.NewRunID()
+	err = g.hooks.PostMergeHook(ctx, HookRecord{
+		EventType:        EventTypePostMerge,
+		RunID:            postRunID,
+		RepositoryID:     repository.RepositoryID,
+		StorageNamespace: storageNamespace,
+		BranchID:         destination,
+
+		SourceRef: commitID.Ref(),
+		Commit:    commit,
+		CommitID:  commitID,
+		PreRunID:  preRunID,
+	})
+	if err != nil {
+		g.log(ctx).
+			WithError(err).
+			WithField("run_id", postRunID).
+			WithField("pre_run_id", preRunID).
+			Error("Post-merge hook failed")
+	}
```

Comment:
looks like we run the post merge hook sync and just log the error.
if we like to block on it, we should return the error or switch the block to work async.

Reply:
post merge hook runs async. **This code was not modified**